### ISSUE TITLE
VMware image stucks at stage 1 waiting for nixos device to appear

### ISSUE THREAD

**Issue by sagikazarmark** [nix-community/nixos-generators#132]
I'm trying to build a vmware image using the example in the [readme](https://github.com/nix-community/nixos-generators#using-in-a-flake) as a starting point.

Unfortunately, booting fails because it cannot find a partition labeled with nixos in stage 1:
<img width="718" alt="Képernyőfotó 2022-01-10 - 11 31 37" src="https://user-images.githubusercontent.com/1226384/148751781-b09b7773-6ee4-4395-a855-d34aef180da9.png">

It might worth adding that I'm using a 21.05 remote builder.

It's absolutely possible that I'm missing something fundamental or that the problem is environmental, I'm still trying to get to the bottom of it. But I thought I'd raise an issue in case it's not or in case someone knows the answer from the top of their head.

**Comment by sagikazarmark** [comment]
Interestingly, importing the VirtualBox format works perfectly.

**Comment by Lassulus** [comment]
hm, interesting. but since we just use https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/vmware-image.nix directly I guess the issue is in nixpkgs. although I'm happy to help you debug if you want to go down that rabbit hole. I guess opening an interactive shell and looking around where the missing partition could be, could be some help. maybe a kernel module is missing or something else

**Comment by Mic92** [comment]
Checkout what device they are using for exposing the block device. Most commonly hypervisors use virtio-blk these days.
Maybe you can find this out in a different linux distribution. Does the generic nixos installer image boots?

**Comment by sagikazarmark** [comment]
Thanks @Mic92 @Lassulus for the quick answer.

Yeah, I agree it's probably a nixpkgs issue, but I have a feeling it's either something very obvious or something hard. If it's obvious, it's probably easier to debug here (at least for me).

Yeah, the generic nixos installer works just fine. Even more interestingly, if I build the vbox image and import it into VMware Fusion, it boots just fine as well.

I opened an interactive shell during one of the attempts and I couldn't see the the device at all. There is a `loading module vmw_pvscsi` line in the screenshot above which I assumed was the right module, but I'll check virtio.

**Comment by sagikazarmark** [comment]
I think I found the missing module. I installed NixOS from using the installer and checked the generated `/etc/nixos/hardware-configuration.nix`.

It contains the following line for me on VMware Fusion:

```
boot.initrd.availableKernelModules = [ "ata_piix" "mptspi" "uhci_hcd" "ehci_pci" "sd_mod" "sr_mod" ];
```

`mptspi` seems to be the missing driver.

Adding it to `boot.initrd.availableKernelModules` resolves the problem.

Now, I don't know whether this is a VMware Fusion only problem, but I have a feeling it is, otherwise I'd expect to see much more issues like this here or in nixpkgs and I couldn't find any.

The above workaround solves the problem for me, but I'd like to report this problem in the appropriate place, I just don't know there. Is it a nixpkgs issue?

**Comment by Lassulus** [comment]
I would add a PR adding the module in this line maybe? could also be it needs another line, but the file seems appropriate to me. 
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/vmware-guest.nix#L30
to test your changes (before commiting), you can do something like this nixos-generate -I nixpkgs=/path/to/your/local/nixpkgs -f ...

**Comment by Mic92** [comment]
Yes. Just add `mptspi` in nixpkgs profile for vmware and post the PR url here, so we can merge it.

**Comment by sagikazarmark** [comment]
PR opened: https://github.com/NixOS/nixpkgs/pull/154307

**Comment by Lassulus** [comment]
so is this still an issue?
